### PR TITLE
feat(document-search): add Ray processing strategy

### PR DESCRIPTION
**Pull request by pawel-chmielak-deepsense**

### CODE REVIEW

**Comment by mhordynski** [comment]
Let's add how-to document about this feature, including how to submit ingestion as a job to a cluster

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            processor_router: The document processor router to use.
+            processor_overwrite: Forces the use of a specific processor, instead of the one provided by the router.
+
+        Returns:
+            A list of elements.
+
+        Raises:
+            ModuleNotFoundError: If Ray is not installed
+        """
+        import ray
```
Comment (kdziedzic68) [inline packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:39]: why not to import at the begining of a file within try-except block? as it is done eg. with litellm or transformers modules within proper classes definitions? anyway we should also log some message like "in order to use blablabla... install the additional dependencies"

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        @ray.remote
+        def process_document_remotely(document: DocumentMeta | Document | Source) -> list[Element]:
+            return asyncio.run(self.process_document(document, processor_router, processor_overwrite))
+
+        tasks = [process_document_remotely.remote(document) for document in documents]
+
+        return sum(await asyncio.gather(*tasks), [])
```
Comment (kdziedzic68) [inline packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:47]: why not to define some batch size?  as with local concurrent processing ?

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+from ragbits.document_search.documents.document import Document, DocumentMeta
+from ragbits.document_search.documents.element import Element
+from ragbits.document_search.documents.sources import Source
+from ragbits.document_search.ingestion.document_processor import DocumentProcessorRouter
+from ragbits.document_search.ingestion.processor_strategies.base import ProcessingExecutionStrategy
+from ragbits.document_search.ingestion.providers.base import BaseProvider
+
+
+class DistributedProcessing(ProcessingExecutionStrategy):
```
[the remark this answers is not in the document]
Comment (pawel-chmielak-deepsense) [reply packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:12]: I'd argue against it - that way we can later swap Ray for sth else without changing the API. Notice the extra requirements group is also called `distributed`, not `ray`.

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py (hunk 2 above)
Comment (pawel-chmielak-deepsense) [reply packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:76]: Good idea, done!

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py (hunk 1 above)
Comment (pawel-chmielak-deepsense) [reply packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py]: Good idea, done!

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdziedzic68) [reply packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:19]: ok - so i thought that maybe we could define some interface for this in that case, but can be left for later

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py (hunk 2 above)
Comment (kdziedzic68) [reply packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:76]: so it is done the way that we have concurent processing in batch and concurrent processing of batches ?

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            processor_router: The document processor router to use.
+            processor_overwrite: Forces the use of a specific processor, instead of the one provided by the router.
+
+        Returns:
+            A list of elements.
+
+        Raises:
+            ModuleNotFoundError: If Ray is not installed
+        """
+        if not HAS_RAY:
```
Comment (kdziedzic68) [inline packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:56]: maybe we should check it while instantiating the object ?

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py (hunk 2 above)
Comment (pawel-chmielak-deepsense) [reply packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py:76]: We split the load into batches and then send batches to different processes. So I'd say it's the latter.

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py (hunk 4 above)
Comment (pawel-chmielak-deepsense) [reply packages/ragbits-document-search/src/ragbits/document_search/ingestion/processor_strategies/distributed.py]: Done

**Review by kdziedzic68 — APPROVED** [review]

**Pull request merged**